pkg/tsdb/phlare/query.go

```diff
@@ -124,114 +127,179 @@ func makeRequest(qm queryModel, query backend.DataQuery) *connect.Request[querie
+			parent = n
+			src = src.Nodes[0]
+			remaining = n.Nodes
+			continue
+		}
+	}
+	if found == nil {
+		if parent == nil {
+			// Nothing in common can't be merged.
+			return
 		}
+		src.Parent = parent
+		parent.Nodes = append(parent.Nodes, src)
+		for p := parent; p != nil; p = p.Parent {
+			p.Value = p.Value + src.Value
+		}
+		return
+	}
+	found.Value = found.Value + src.Self
+	for p := found.Parent; p != nil; p = p.Parent {
+		p.Value = p.Value + src.Self
+	}
+	found.Self = found.Self + src.Self
+}
 
-		var nextParentsStack []*ProfileTree
-		currentParent := parentsStack[:1][0]
-		parentsStack = parentsStack[1:]
-		itemIndex := 0
-		// cumulative offset as items in flamebearer format have just relative to prev item
-		offset := int64(0)
+func treeFromSample(profile *googlev1.Profile, sample *googlev1.Sample) *ProfileTree {
+	if len(sample.LocationId) == 0 {
+		return &ProfileTree{
+			Level: 0,
+			Value: sample.Value[0],
+			Function: &Function{
+				FunctionName: "root",
+			},
+		}
+	}
 
-		// Cycle through bar in a level
-		for {
-			if itemIndex >= len(levels[currentLevel].Values) {
-				break
-			}
+	// The leaf is at locations[0].
+	locations := sample.LocationId
 
-			itemStart := levels[currentLevel].Values[itemIndex+START_OFFSET] + offset
-			itemValue := levels[currentLevel].Values[itemIndex+VALUE_OFFSET]
-			selfValue := levels[currentLevel].Values[itemIndex+SELF_OFFSET]
-			itemEnd := itemStart + itemValue
-			parentEnd := currentParent.Start + currentParent.Value
-
-			if itemStart >= currentParent.Start && itemEnd <= parentEnd {
-				// We have an item that is in the bounds of current parent item, so it should be its child
-				treeItem := &ProfileTree{
-					Start: itemStart,
-					Value: itemValue,
-					Self:  selfValue,
-					Level: currentLevel,
-					Name:  names[levels[currentLevel].Values[itemIndex+NAME_OFFSET]],
-				}
-				// Add to parent
-				currentParent.Nodes = append(currentParent.Nodes, treeItem)
-				// Add this item as parent for the next level
-				nextParentsStack = append(nextParentsStack, treeItem)
-				itemIndex += ITEM_OFFSET
-
-				// Update offset for next item. This is changing relative offset to absolute one.
-				offset = itemEnd
-			} else {
-				// We went out of parents bounds so lets move to next parent. We will evaluate the same item again, but
-				// we will check if it is a child of the next parent item in line.
-				if len(parentsStack) == 0 {
-					logger.Error("parentsStack is empty but there are still items in current level", "currentLevel", currentLevel, "itemIndex", itemIndex)
-					break
-				}
-				currentParent = parentsStack[:1][0]
-				parentsStack = parentsStack[1:]
-				continue
-			}
+	current := &ProfileTree{
+		Self:  sample.Value[0],
+		Level: 0,
+	}
+	for len(locations) > 0 {
+		current.locationID = locations[0]
+		current.Value = sample.Value[0]
+		current.Level = len(locations)
+
+		lines := profile.Location[locations[0]-1].Line
```

Comment:
Why do we adjust the index to `locations[0]-1`? On previous line we set `current.locationID = locations[0]` but then seem to add lines from parent location?